### PR TITLE
update NecroValley

### PR DESCRIPTION
**Pull request by salix5**
@mercury233 

# Problem
#1860 

# Solution
`EFFECT_NECRO_VALLEY_CHECK`
Check if the activating card is affected by NecroValley.

# Reference
https://ocg-rule.readthedocs.io/zh_CN/latest/c04/%E7%8E%8B%E5%AE%B6%E9%95%BF%E7%9C%A0%E4%B9%8B%E8%B0%B7.html#id4
> 而不受魔法效果影响的怪兽，发动的效果不管是否只涉及墓地，都不受「王家长眠之谷」的②效果影响，仍会正常适用。

### CODE REVIEW

**Comment by mercury233** [comment]
confirming

File: c83764718.lua [hunk 1]
```diff
@@ -23,7 +23,7 @@ function c83764718.target(e,tp,eg,ep,ev,re,r,rp,chk,chkc)
 end
 function c83764718.activate(e,tp,eg,ep,ev,re,r,rp)
 	local tc=Duel.GetFirstTarget()
-	if tc:IsRelateToEffect(e) and aux.NecroValleyFilter()(tc) then
+	if tc:IsRelateToEffect(e) and e:GetHandler():IsHasEffect(EFFECT_NECRO_VALLEY_CHECK) and aux.NecroValleyFilter()(tc) then
```
Comment (mercury233) [inline c83764718.lua:26]: _and e:GetHandler() is related to e_
Maybe getting the current effect handler and NecroValley player effect in `NecroValleyFilter` and check immunity will work? I know card immune to player effect can be weird.

Comment (salix5) [reply c83764718.lua:26]: EFFECT_NECRO_VALLEY
for monsters: The monster is protected by Necrovalley.
for players: The Graveyard of the player is protected by Necrovalley.

I think this check should use a different effect code.

**Comment by salix5** [comment]
@mercury233 
Can we merge this one?

**Comment by mercury233** [comment]
I don't think so but I don't know why

**Comment by salix5** [comment]
Auxiliary.NecroValleyFilter()
Auxiliary.NecroValleyNegateCheck()
這些函數的檢查都會使用IsHasEffect(EFFECT_NECRO_VALLEY)

此處IsHasEffect(EFFECT_NECRO_VALLEY)的定義應該是：
這張卡在王谷保護的墓地中

如果我們要檢查【這張卡是否會被王谷的效果無效】
我們可能應該採用其他的效果避免混淆
例如這個PR提出的EFFECT_NECRO_VALLEY_CHECK

**Pull request closed without merging**